run_project_tests.py

```diff
@@ -1045,61 +1074,103 @@ def __init__(self, category: str, subdir: str, skip: bool = False, stdout_mandat
     return gathered_tests
 
 def run_tests(all_tests: T.List[T.Tuple[str, T.List[TestDef], bool]],
-              log_name_base: str, failfast: bool,
-              extra_args: T.List[str], use_tmp: bool) -> T.Tuple[int, int, int]:
-    global logfile
+              log_name_base: str,
+              failfast: bool,
+              extra_args: T.List[str],
+              use_tmp: bool,
+              num_workers: int) -> T.Tuple[int, int, int]:
     txtname = log_name_base + '.txt'
     with open(txtname, 'w', encoding='utf-8', errors='ignore') as lf:
-        logfile = lf
-        return _run_tests(all_tests, log_name_base, failfast, extra_args, use_tmp)
+        return _run_tests(all_tests, log_name_base, failfast, extra_args, use_tmp, num_workers, lf)
+
+class TestStatus(Enum):
+    OK       = normal_green(' [SUCCESS] ')
+    SKIP     =       yellow(' [SKIPPED] ')
+    ERROR    =          red('  [ERROR]  ')
+    CANCELED =         cyan('[CANCELED] ')
+    RUNNING  =         blue(' [RUNNING] ')  # Should never be actually printed
+    LOG      =         bold('   [LOG]   ')  # Should never be actually printed
+
+def default_print(*args: mlog.TV_Loggable, sep: str = ' ') -> None:
+    print(*args, sep=sep)
+
+safe_print = default_print
+
+class TestRunFuture:
+    def __init__(self, name: str, testdef: TestDef, future: Future[T.Optional[TestResult]]) -> None:
+        super().__init__()
+        self.name = name
+        self.testdef = testdef
+        self.future = future
+        self.status = TestStatus.RUNNING
+
+    @property
+    def result(self) -> T.Optional[TestResult]:
+        return self.future.result()
+
+    def log(self) -> None:
+        without_install = "" if len(install_commands) > 0 else "(without install)"
```

Comment:
this should be just `'' if install_commands else "(without install)"`, using `len()` to check container emptiness is a python anti-pattern.

Reply:
Yeah, missed that when I copy-pasted code around. Is fixed now.